### PR TITLE
Series `apply` method backed by masked UDFs

### PR DESCRIPTION
**Pull request by brandon-b-miller**
Depends on https://github.com/rapidsai/cudf/pull/9174

Adds `Series.apply` which applies a scalar UDF elementwise to the series data returning a new series. Null sensitive. Works in terms of our numba `MaskedType` extension type. Similar to `pd.Series.apply`.

### CODE REVIEW

File: python/cudf/cudf/core/udf/api.py [hunk 1]
```diff
@@ -14,3 +14,11 @@ class Masked:
     def __init__(self, value, valid):
         self.value = value
         self.valid = valid
+
+
+def pack_return(masked_or_scalar):
+    # Blank function to give us something for the typing and
+    # lowering to grab onto. Just a dummy function for us to
```
Comment (shwina) [inline python/cudf/cudf/core/udf/api.py:21]:
```suggestion
    # lowering to grab onto. Just a placeholder function for us to
```

https://developers.google.com/style/inclusive-documentation

File: python/cudf/cudf/core/udf/pipeline.py [hunk 2]
```diff
@@ -50,3 +64,155 @@ def wrapper(*args):
         return to_udf_table._apply(func)
 
     return wrapper
+
+
+def masked_arrty_from_np_type(dtype):
```
Comment (thomcom) [inline python/cudf/cudf/core/udf/pipeline.py:69]:
```suggestion
def masked_array_from_np_type(dtype):
```

File: python/cudf/cudf/core/udf/pipeline.py [hunk 3]
```diff
@@ -50,3 +64,155 @@ def wrapper(*args):
...
+    actually JIT the kernel itself later, accounting for types
+    and offsets
+    """
+
+    # Tuple of arrays, first the output data array, then the mask
+    return_type = Tuple((return_type[::1], boolean[::1]))
+    offsets = []
+    sig = [return_type]
+    for col in df._data.values():
+        sig.append(masked_arrty_from_np_type(col.dtype))
```
Comment (thomcom) [inline python/cudf/cudf/core/udf/pipeline.py:92]:
```suggestion
        sig.append(masked_array_from_np_type(col.dtype))
```

File: python/cudf/cudf/core/udf/pipeline.py [hunk 4]
```diff
@@ -50,3 +64,155 @@ def wrapper(*args):
...
+        offsets.append(int64)
```
Comment (thomcom) [inline python/cudf/cudf/core/udf/pipeline.py:93]: I'm curious about `offsets` being fixed to 64 bit? Is this a limitation of numba or something you're going to add?

File: python/cudf/cudf/core/udf/pipeline.py [hunk 5]
```diff
@@ -50,3 +64,155 @@ def wrapper(*args):
...
+
+    # return_type + data,masks + offsets + size
+    sig = void(*(sig + offsets + [int64]))
```
Comment (thomcom) [inline python/cudf/cudf/core/udf/pipeline.py:96]: I do not like using the same `int64` for values inserted into `offsets` and also for the argument to `_kernel`.

File: python/cudf/cudf/core/udf/pipeline.py [hunk 6]
```diff
@@ -50,3 +64,155 @@ def wrapper(*args):
...
+
+    # Incorporate all of the above into the kernel code template
+    d = {
+        "input_columns": input_columns,
+        "input_offsets": input_offsets,
+        "masked_input_initializers": masked_input_initializers,
+        "user_udf_call": user_udf_call,
+    }
+
+    return kernel_template.format(**d)
```
Comment (thomcom) [inline python/cudf/cudf/core/udf/pipeline.py:164]: fun!

File: python/cudf/cudf/core/udf/pipeline.py [hunk 7]
```diff
@@ -50,3 +64,155 @@ def wrapper(*args):
...
+            # Defines a kernel named "_kernel" in the lcl dict
+            _define_function(df, scalar_return=_is_scalar_return),
+            {
+                "f_": f_,
+                "cuda": cuda,
+                "Masked": Masked,
+                "mask_get": mask_get,
+                "pack_return": pack_return,
+            },
+            lcl,
```
Comment (thomcom) [inline python/cudf/cudf/core/udf/pipeline.py:211]: What is an `lcl dict`?

**Review by thomcom — CHANGES_REQUESTED** [review]
Very cool contribution. Just a few bits in commentary.

**Comment by brandon-b-miller** [comment]
@thomcom thank you for the review! would you mind moving your comments to https://github.com/rapidsai/cudf/pull/9174, this PR is just a minor set of diffs on top of that one. Hopefully I can address them there!

**Comment by brandon-b-miller** [comment]
rerun tests

**Comment by brandon-b-miller** [comment]
This is now ready for review.

**Review by shwina — APPROVED** [review]
This looks great to me!

**Review by thomcom — APPROVED** [review]

**Comment by brandon-b-miller** [comment]
rerun tests

**Review by galipremsagar — APPROVED** [review]

**Comment by brandon-b-miller** [comment]
@gpucibot merge

**Pull request merged**